builtin/credential/aws/path_login.go

```diff
@@ -0,0 +1,492 @@
+package aws
+
+import (
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+	"github.com/vishalnayak/pkcs7"
+)
+
+func pathLogin(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "login$",
+		Fields: map[string]*framework.FieldSchema{
+			"pkcs7": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "PKCS7 signature of the identity document.",
+			},
+
+			"nonce": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "The nonce created by a client of this backend.",
+			},
+		},
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLoginUpdate,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+// validateInstanceID queries the status of the EC2 instance using AWS EC2 API and
+// checks if the instance is running and is healthy.
+func validateInstanceID(s logical.Storage, instanceID string) error {
+	// Create an EC2 client to pull the instance information
+	ec2Client, err := clientEC2(s)
+	if err != nil {
+		return err
+	}
+
+	// Get the status of the instance
+	instanceStatus, err := ec2Client.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+		InstanceIds: []*string{aws.String(instanceID)},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Validate the instance through InstanceState, InstanceStatus and SystemStatus
+	return validateInstanceStatus(instanceStatus)
+}
+
+// validateMetadata matches the given client nonce and pending time with the one cached
+// in the identity whitelist during the previous login.
+func validateMetadata(clientNonce, pendingTime string, storedIdentity *whitelistIdentity, imageEntry *awsImageEntry) error {
+
+	givenPendingTime, err := time.Parse(time.RFC3339, pendingTime)
+	if err != nil {
+		return err
+	}
+
+	storedPendingTime, err := time.Parse(time.RFC3339, storedIdentity.PendingTime)
+	if err != nil {
+		return err
+	}
+
+	// When the presented client nonce does not match the cached entry, it is
+	// either that a rogue client is trying to login or that a valid client
+	// suffered a migration. The migration is detected via pendingTime in the
```

Comment:
The nonce is specific to this backend because of the TOFU property of the credential coming from Amazon. Other credential backends don't have this property.

If you're having trouble figuring out how to rotate specific tokens or secrets, feel free to post on the mailing list and I can help you out.